blockchain/vm/contracts.go

```diff
@@ -56,40 +56,38 @@ type PrecompiledContract interface {
 	GetRequiredGasAndComputationCost(input []byte) (uint64, uint64)
 
 	// Run runs the precompiled contract
-	Run(input []byte) ([]byte, error)
+	// contract, evm is only exists in klaytn, those are not used in go-ethereum
+	Run(input []byte, contract *Contract, evm *EVM) ([]byte, error)
 }
 
-// vmLogAddress is the address of precompiled contract vmLog.
-var vmLogAddress = common.BytesToAddress([]byte{9})
-
-// feePayerAddress is the address of precompiled contract feePayer.
-var feePayerAddress = common.BytesToAddress([]byte{10})
-
-// validateSenderAddress is the address of precompiled contract ValidateSender.
-var validateSenderAddress = common.BytesToAddress([]byte{11})
-
-// PrecompiledContractsCypress contains the default set of pre-compiled contracts.
+// PrecompiledContractsCypress contains the default set of pre-compiled Klaytn
+// contracts used in the Cypress release.
 var PrecompiledContractsCypress = map[common.Address]PrecompiledContract{
-	common.BytesToAddress([]byte{1}): &ecrecover{},
-	common.BytesToAddress([]byte{2}): &sha256hash{},
-	common.BytesToAddress([]byte{3}): &ripemd160hash{},
-	common.BytesToAddress([]byte{4}): &dataCopy{},
-	common.BytesToAddress([]byte{5}): &bigModExp{},
-	common.BytesToAddress([]byte{6}): &bn256Add{},
-	common.BytesToAddress([]byte{7}): &bn256ScalarMul{},
-	common.BytesToAddress([]byte{8}): &bn256Pairing{},
-	vmLogAddress:                     &vmLog{},
-	feePayerAddress:                  &feePayer{},
-	validateSenderAddress:            &precompiledValidateSender{},
+	common.BytesToAddress([]byte{1}):  &ecrecover{},
+	common.BytesToAddress([]byte{2}):  &sha256hash{},
+	common.BytesToAddress([]byte{3}):  &ripemd160hash{},
+	common.BytesToAddress([]byte{4}):  &dataCopy{},
+	common.BytesToAddress([]byte{5}):  &bigModExp{},
+	common.BytesToAddress([]byte{6}):  &bn256Add{},
+	common.BytesToAddress([]byte{7}):  &bn256ScalarMul{},
+	common.BytesToAddress([]byte{8}):  &bn256Pairing{},
+	common.BytesToAddress([]byte{9}):  &vmLog{},
+	common.BytesToAddress([]byte{10}): &feePayer{},
+	common.BytesToAddress([]byte{11}): &validateSender{},
 }
 
+// TODO-IncompatibleChange: define PrecompiledContracts set for first incompatible change
+
 // RunPrecompiledContract runs and evaluates the output of a precompiled contract.
-func RunPrecompiledContract(p PrecompiledContract, input []byte, contract *Contract) (ret []byte, computationCost uint64, err error) {
+func RunPrecompiledContract(p PrecompiledContract, input []byte, contract *Contract, evm *EVM) (ret []byte, computationCost uint64, err error) {
 	gas, computationCost := p.GetRequiredGasAndComputationCost(input)
 	if contract.UseGas(gas) {
-		ret, err := p.Run(input)
+		ret, err = p.Run(input, contract, evm)
 		return ret, computationCost, err
 	}
+	if p == (&validateSender{}) {
```

Comment:
This is always false because `&validateSender{}` always creates a new struct. Checkout below :
```
	a := &validateSender{}
	b := &validateSender{}
	t.Log(a == b) // false
```
How about implementing like this? : 
```
	if _, ok := p.(*validateSender); ok {
```

Reply:
Thanks, it worked well!! However, according to below comment, i rechecked whether this logic is necessary or not. In conclusion, it is not necessary, so it will be deleted!